Use Python for the local building script

stevekuznetsov opened the pull request:
We have a complicated relationship to describe between the canonical
names of the images we need to build, their usual source context
directories, files in those directories that we would want to update in
the containers, and binaries from the outut directory that those images
depend on. We are much better off writing the script to build these
images using Python than Bash, as modeling those relationships is easier
with a language that has nesting associative arrays.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>


/cc @bparees

bparees commented on hack/lib/util/misc.sh:
```diff
@@ -181,7 +181,7 @@ function os::util::curl_etcd() {
 			               -password "pass:${etcd_client_cert_p12_password}"
 		fi
-
+b
```
typo?

bparees commented on hack/lib/util/misc.sh:
```diff
@@ -204,3 +204,31 @@ function os::util::host_platform() {
 }
 readonly -f os::util::host_platform
+
+# os::util::list_go_src_files lists files we consider part of our project
```
does not appear to be an intended part of this pr.

stevekuznetsov commented:
@bparees updated

bparees commented:
all my beautiful bash code.......

stevekuznetsov commented:
/cc @soltysh

bparees commented on hack/build-local-images.py:
```diff
@@ -0,0 +1,136 @@
+		},
+		"files": {}
+	},
+	"openshift/origin-deployer": {
```
you can't hardcode the origin prefix, this should work when i'm using it with ocp too.

bparees commented on hack/build-local-images.py:
```diff
@@ -0,0 +1,136 @@
+
+	debug("Initiating Docker build with Dockerfile:\n{}".format(open(join(context_dir, "Dockerfile")).read()))
+	call(["docker", "build", "-t", image, "."], cwd=context_dir)
+	rmtree(context_dir)
```
you're rebuilding and destroying the context dir for every image?  that means you're going to copy the 300meg openshift binary 8 times if you're building all the images.  This is the kind of thing my bash script was trying to avoid.

stevekuznetsov commented:
@bparees 


Addressed comments:

 - allowed for image namespace and prefix configuration using
   `${OS_IMAGE_PREFIX}` as in the Bash lib
 - re-use one temporary context dir for all builds, copy the
   binaries only once

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

bparees commented:
well it lgtm but i'm still suspecting this is heading down a path @smarterclayton doesn't want to go in terms of effectively recreating the Dockerfiles in the form of a python map.

It's also going to be still slower than the existing function that uses the openshift binary in place instead of copying it, but given that you only copy it once and then reuse it, that's probably ok.  (obviously the existing script is not in any way extensible to use content beyond what's in the binary dir).

My only other worry w/ the contextdir reuse is that docker "uploads" the entire content of the contextdir to the daemon on a dockerbuild.  So by the time you get to building the last image, you may be uploading a ton of content (that you don't really need) as you accumulate junk from earlier images.  Again, not a problem today since all the images have only a single file and it's common between them all.

So in sum, i'm in favor of merging it, but i do still see potential growing pains (vs the existing script which basically wasn't designed to grow :) )

stevekuznetsov commented:
> effectively recreating the Dockerfiles in the form of a python map.

ack

> as you accumulate junk from earlier images. 

can prune `<context>/src/<image>` after build is finished

smarterclayton commented:
TBH this about 10 times more work into this than it deserved.  It's a hack script, it doesn't have to be pretty, well factored, or complex.  It just needs to work and for us to keep it small.  I don't care that it is python, I just want something people can run.

These sorts of scripts should be deliberately simple and ugly to force people to come up with a Real Solution, which we are already discussing.  I'm fine with it, let's not paint the shed any further though.

bparees commented:
@smarterclayton yeah that's kinda what i figured (and i don't disagree)
@stevekuznetsov squash and i'll merge.

smarterclayton commented:
The thing to optimize for is "how long does it take anyone on the team to understand to where they can add a new file into the image for easy iteration for their own purpose."  Its mostly achieved here, but I'd be biased towards simpler and uglier if it meant people didn't have to waste too much time getting changes here.  This isn't bad, just trying to keep this about making it easy for anyone to consume.  Steve should never have to touch this file, but folks on Derek's team will.

stevekuznetsov commented:
Squashed

bparees commented:
[merge]
